### PR TITLE
Implement majority of more efficient Redshift Copy Strategy.

### PR DESCRIPTION
**Pull request by davinchia**
## What
Solves the redshift portion of #1958.

Technical approach discussed [here](https://docs.google.com/document/d/1BeDx5v4t9JcO2PVUkV7iXlxgro3AuTc4JrnUiqssHfg/edit#heading=h.gn4ad5et1brq).

## How
TLDR:
Instead of inserts, we write the data to S3 and issue a COPY command copying from S3 into redshift.

- Use a single file as its sufficiently performant and we do not want to introduce file destination related operations yet.
- Use an open source library for uploads as AWS does not natively support streaming loads.

See the linked doc for more detail.

My intention with this PR is to first implement the meat of the write-and-copy strategy. This is mainly centered around the `RedshiftCopier` class. I plan hook up the `RedshiftCopier` to the actual Destination class, and implement all the plumbing, in a follow up PR.

## Pre-merge Checklist
- [ ] *Finish speed tests.*

## Recommended reading order
1. `RedshiftCopier`
2. `RedshiftCopyDestination`

### CODE REVIEW

File: airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftCopierTest.java
Comment (davinchia) [inline airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftCopierTest.java:55]: will switch this to an env file before merging.

File: airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftCopierTest.java
Comment (davinchia) [inline airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftCopierTest.java:66]: even though `getRawTableName` is deprecated - it's less confusing to use the deprecated method for now, and refactor everything at one go later. I'll put up a PR once everything is done.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/DestinationConsumer.java
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/DestinationConsumer.java:41]: Nice, too bad the spotlessStyling is messing up the list view...

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java
Comment (cgardens) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java:154]: is there a way to do this in a single transaction for all tables? if not then we need to document expectations in the error cases. currently for other destinations, we are able to do this copy atomically.

**Review by cgardens — CHANGES_REQUESTED** [review]
the approach makes sense. is the goal to merge this and then do a separate PR or to keep this open until your second PR is done? the bones of this seem right, one concern about the atomicity of the copy operation. after that's resolved i think we will be good.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftSqlOperations.java
Comment (davinchia) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftSqlOperations.java:86]: You are right. The more recommended way is to use roles. Will change.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopyDestination.java
Comment (davinchia) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopyDestination.java:124]: Yes it should. I've left it out for now; planned for these changes to be part of the follow up PR.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java
Comment (davinchia) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java:115]: Yes if we set too high of a queue * part limit combination. As of now, our memory consumption should 200 MB (100MB sitting in queue, 100MB uploading), which seems okay for now? I can leave a comment warning devs off setting this too high?

File: airbyte-integrations/connectors/destination-redshift/build.gradle
Comment (davinchia) [reply airbyte-integrations/connectors/destination-redshift/build.gradle:33]: haha I was talking to someone and didn't realise we have a convention. Will chat with you about this during our sync later.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java
Comment (davinchia) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java:154]: great point. nothing jumps out at me for now; will play around and change this.

**Comment by davinchia** [comment]
@cgardens thanks for the quick turnaround. plan is to merge this in first and open another PR. want to not leave hanging PRs and I myself don't like merging in massive blocks of code. lmk if we generally don't favor that. will work on the feedback now

**Review by cgardens — APPROVED** [review]
@davinchia and I discussed the transaction piece. Switching to approve because we are the same page on what needs to be done and I don't block this PR due to timezone concerns.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java
Comment (davinchia) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java:140]: @cgardens this was the cleanest way I could find of doing this - basically the same pattern as the `JdbcBufferedConsumerFactory`. If I understand correctly, the `executeTransaction` call on L145 will automatically rollback any inserts into the final destination tables if inserts run into any errors.

The tmp table and the S3 files will be cleaned up by the finally block.

**Pull request merged**
